File: modules/student-experience/src/main/java/com/campusos/portal/payload/DocumentPayload.java
package com.campusos.portal.payload;

import com.campusos.portal.domain.DocType;
import com.campusos.portal.domain.RequestType;
import java.util.ArrayList;
import java.util.List;

public class DocumentPayload implements RequestPayload {

    /* ---- frozen contract fields ---- */
    public DocType docType;
    public String purpose;
    public int copies = 1;

    /* ---- system-derived ---- */
    public Sys sys = new Sys();

    public static class Sys {
        public Boolean autoEligible;
        public String eligibilityReason;
        public String serialNo;
        public String verifyId;
        public Long documentId;
    }

    @Override public RequestType type() { return RequestType.DOCUMENT; }

    @Override public String title() { return docType.display(); }

    @Override public String subtitle() {
        return copies + " copy/copies · " + purpose;
    }

    @Override public List<Artifact> artifacts() {
        List<Artifact> out = new ArrayList<>();
        if (sys.serialNo != null) out.add(Artifact.of("SERIAL", "Serial number", sys.serialNo));
        if (sys.verifyId != null) {
            out.add(new Artifact("VERIFY_ID", "Verification ID", sys.verifyId, "/verify/" + sys.verifyId));
        }
        if (sys.documentId != null) {
            out.add(new Artifact("DOCUMENT", "Generated document", "View / download",
                    "/documents/" + sys.documentId + "/download"));
        }
        return out;
    }

    @Override public void validate() {
        if (docType == null) throw new IllegalArgumentException("docType is required");
        if (purpose == null || purpose.isBlank()) throw new IllegalArgumentException("purpose is required");
        if (copies < 1 || copies > 3) throw new IllegalArgumentException("copies must be 1–3");
    }
}

File: modules/student-experience/src/payload/DocumentPayload.js
import { RequestType, DocType } from '../domain/enums.js';
import { IllegalArgumentException } from './RequestPayload.js';
import { artifact } from './Artifact.js';
import { I18n } from '../view/i18n.js';

export class DocumentPayload {
  constructor() {
    this.docType = null;
    this.purpose = null;
    this.copies = 1;
    this.sys = { autoEligible: null, eligibilityReason: null, serialNo: null, verifyId: null, documentId: null };
  }

  type() { return RequestType.DOCUMENT; }

  title(locale = 'en') { return DocType.display(this.docType, locale); }

  subtitle(locale = 'en') { return `${this.copies} ${I18n.t(locale, 'payload.document.copySuffix')} · ${this.purpose}`; }

  artifacts(locale = 'en') {
    const out = [];
    if (this.sys.serialNo !== null) out.push(artifact('SERIAL', I18n.t(locale, 'artifact.serialNumber'), this.sys.serialNo));
    if (this.sys.verifyId !== null) {
      out.push(artifact('VERIFY_ID', I18n.t(locale, 'artifact.verificationId'), this.sys.verifyId, `/verify/${this.sys.verifyId}`));
    }
    if (this.sys.documentId !== null) {
      out.push(artifact('DOCUMENT', I18n.t(locale, 'artifact.generatedDocument'), I18n.t(locale, 'artifact.viewDownload'), `/documents/${this.sys.documentId}/download`));
    }
    return out;
  }

  handledBy() { return null; }

  validate() {
    if (!this.docType || !DocType.values().includes(this.docType)) {
      throw new IllegalArgumentException('docType is required');
    }
    if (!this.purpose || !this.purpose.trim()) throw new IllegalArgumentException('purpose is required');
    // Number.isInteger(NaN) is false — this is the gate every payload passes through, so a
    // non-numeric `copies` must be caught here even though the form layer also checks it.
    if (!Number.isInteger(this.copies) || this.copies < 1 || this.copies > 3) {
      throw new IllegalArgumentException('copies must be 1–3');
    }
  }

  static fromJSON(o) { return Object.assign(new DocumentPayload(), o, { sys: { ...new DocumentPayload().sys, ...o.sys } }); }
}
